src/ruby/pb/test/xds_client.rb

```diff
@@ -0,0 +1,193 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This is the xDS interop test Ruby client. This is meant to be run by
+# the run_xds_tests.py test runner.
+#
+# Usage: $ tools/run_tests/run_xds_tests.py --test_case=... ...
+#    --client_cmd="path/to/xds_client.rb --server=<hostname> \
+#                                        --stats_port=<port> \
+#                                        --qps=<qps>"
+
+# These lines are required for the generated files to load grpc
+this_dir = File.expand_path(File.dirname(__FILE__))
+lib_dir = File.join(File.dirname(File.dirname(this_dir)), 'lib')
+pb_dir = File.dirname(this_dir)
+$LOAD_PATH.unshift(lib_dir) unless $LOAD_PATH.include?(lib_dir)
+$LOAD_PATH.unshift(pb_dir) unless $LOAD_PATH.include?(pb_dir)
+
+require 'optparse'
+require 'logger'
+
+require_relative '../../lib/grpc'
+require 'google/protobuf'
+
+require_relative '../src/proto/grpc/testing/empty_pb'
+require_relative '../src/proto/grpc/testing/messages_pb'
+require_relative '../src/proto/grpc/testing/test_services_pb'
+
+# Some global variables to be shared by server and client
+$watchers = Array.new
+$watchers_mutex = Mutex.new
+$current_request_id = 0;
+
+# RubyLogger defines a logger for gRPC based on the standard ruby logger.
+module RubyLogger
+  def logger
+    LOGGER
+  end
+
+  LOGGER = Logger.new(STDOUT)
+  LOGGER.level = Logger::INFO
+end
+
+# GRPC is the general RPC module
+module GRPC
+  # Inject the noop #logger if no module-level logger method has been injected.
+  extend RubyLogger
+end
+
+# creates a test stub
+def create_stub(opts)
+  address = "#{opts.server}"
+
+  GRPC.logger.info("... connecting insecurely to #{address}")
+  Grpc::Testing::TestService::Stub.new(
+    address,
+    :this_channel_is_insecure,
+  )
+end
+
+# This implements LoadBalancerStatsService required by the test runner
+class TestTarget < Grpc::Testing::LoadBalancerStatsService::Service
+  include Grpc::Testing
+
+  def get_client_stats(req, _call)
+    start_id = $current_request_id + 1
+    end_id = start_id + req['num_rpcs'] # watch the next X rpcs
+    start_time = Time.now.to_f
+    finish_time = start_time + req['timeout_sec']
+    index = 0
+    $watchers_mutex.synchronize do
+      index = $watchers.length
+      # Add a watcher for this range of id's
+      $watchers << {
+        "start_id" => start_id,
+        "end_id" => end_id,
+        "rpcs_by_peer" => Hash.new(0),
+        "rpcs_needed" => req['num_rpcs'],
+        "no_remote_peer" => 0
+      }
+    end
+    # wait until all RPCs are done, or we are past timeout
+    while ($watchers[index]['rpcs_needed'] > 0) &&
```

Comment:
we can use a `ConditionVariable` to replace the sleeps here.

Let's add a new global, `$watchers_cv = ConditionVariable.new`

Then do something like this (this also removes `watcher` after it's no longer used)

```
$watchers_mutex.synchronize do
  watcher = { ... }
  $watchers << watcher
  seconds_remaining = Process.clock_gettime(Process::CLOCK_MONOTONIC) - finish_time
  while watcher['rpcs_needed'] > 0 && seconds_remaining > 0
    $watchers_cv.wait($watchers_mutex, seconds_remaining)
    seconds_remaining = Process.clock_gettime(Process::CLOCK_MONOTONIC) - finish_time
  end
  $watchers.delete_at($watchers.index watcher)
end
LoadBalancerStatsResponse.new(...)
```

Reply:
Fixed. Thanks for pointing me to `ConditionVariable`.